src/extensions/house/houseext_hooks.cpp

```diff
@@ -154,6 +159,128 @@ DECLARE_PATCH(_HouseClass_Can_Build_BuildCheat_Patch)
+        return count;
+    }
+
+    /**
+     *  #issue-611
+     *
+     *  If the object has a build limit, then reduce count by 1
+     */
+    if (technotype->BuildLimit > 0) {
+        count--;
+    }
+
+    /**
+    *  #issue-715
+    *
+    *  If the object can transform into another object through our special logic,
+    *  then check that doing so doesn't allow circumventing build limits
+    */
+    if (technotype->What_Am_I() == RTTI_UNITTYPE) {
+        UnitTypeClass* unittype = reinterpret_cast<UnitTypeClass*>(technotype);
+        UnitTypeClassExtension* unittypeext = Extension::Fetch<UnitTypeClassExtension>(unittype);
+
+        if (unittype->DeploysInto == nullptr && unittypeext->TransformsInto != nullptr) {
+            count += factory->House->UQuantity.Count_Of((UnitType)(unittypeext->TransformsInto->Get_Heap_ID()));
+        }
+    }
+
+    return count;
+}
+
+
+/**
+ *  #issue-611, #issue-715
+ *
+ *  Fixes the game allowing the player to queue one unit too few
+ *  when a unit has BuildLimit > 1.
+ *
+ *  Also updates the logic to take our unit transformation logic
+ *  into account.
+ */
+DECLARE_PATCH(_HouseClass_ShouldDisableCameo_BuildLimit_Fix)
+{
+    GET_REGISTER_STATIC(FactoryClass*, factory, ecx);
+    GET_REGISTER_STATIC(TechnoTypeClass*, technotype, esi);
+    static int queuedcount;
+
+    queuedcount = _HouseClass_ShouldDisableCameo_Get_Queued_Count(factory, technotype);
+
+    _asm { mov eax, [queuedcount] }
+    JMP_REG(ecx, 0x004CB77D);
+}
+
+
+/**
+ *  #issue-715
+ *
+ *  Take vehicles that can transform into other vehicles into acccount when
+ *  determining whether a build limit has been met/exceeded.
+ *  Otherwise these kinds of units could be used to bypass build limits
+ *  (build a limited vehicle, transform it, now you can build another vehicle).
+ *
+ *  Author: Rampastring
+ */
+DECLARE_PATCH(_HouseClass_Can_Build_BuildLimit_Handle_Vehicle_Transform)
+{
+    GET_REGISTER_STATIC(UnitTypeClass*, unittype, edi);
+    GET_REGISTER_STATIC(HouseClass*, house, ebp);
+    static UnitTypeClassExtension* unittypeext;
+    static int objectcount;
+
+    unittypeext = Extension::Fetch<UnitTypeClassExtension>(unittype);
+
+    /**
+     *  Stolen bytes / code.
+     */
+    objectcount = house->UQuantity.Count_Of((UnitType)unittype->Get_Heap_ID());
+
+    /**
+     *  Check whether this unit can deploy into a building.
+     *  If it can, increment the object count by the number of buildings.
+     */
+    if (unittype->DeploysInto != nullptr) {
+        objectcount += house->BQuantity.Count_Of((BuildingType)unittype->DeploysInto->Get_Heap_ID());
+    }
+    else if (unittypeext->TransformsInto != nullptr) {
+
+        /**
+         *  This unit can transform into another unit, increment the object count
+         *  by the number of transformed units.
+         */
+        objectcount += house->UQuantity.Count_Of((UnitType)(unittypeext->TransformsInto->Get_Heap_ID()));
```

Comment:
Be careful of such a large operation here. Check the assembly, hopefully, this only uses EAX for each operation.